### ISSUE TITLE
svd stuck in infinite loop for certain matrices

### ISSUE THREAD

**Issue by cvsvensson** [JuliaLinearAlgebra/GenericLinearAlgebra.jl#81]
Using svd on certain matrices leads to an infinite loop:
```julia
using LinearAlgebra, GenericLinearAlgebra
m = [1 0 0 0; 0 2 1 0; 0 1 2 0; 0 0 0 -1]
svd(m) # ok
svd(BigFloat.(m)) # stuck in infinite loop
svdvals!(Float64.(m), debug=true) # also stuck, so it's not BigFloat's fault.
```
The svd of some permutations of the matrix works:
```julia
perm = [2,3,1,4]
svd(BigFloat.(m[perm,perm])) # ok
```
For this particular matrix, any small perturbation (e.g. `m + 1e-10*rand(4,4)`) seem to work fine but there are other matrices, such as 
```julia
m2 = [0.4 0 0 0 0 0 0 0; 0 0.8 0 0 0 0 1 0; 0 0 -0.3 0 0 0 0 0; 0 0 0 -0.4 0 0 0 0; 0 0 0 0 0.4 0 0 0; 0 0 0 0 0 -0.3 0 0; 0 1 0 0 0 0 0.8 0; 0 0 0 0 0 0 0 -0.4]
```
where `svdvals!(m2+1e-2*rand(8,8), debug=true)` often gets stuck.

The problem may be related to the code linked below. If I replace `shift^2 < eps(B.dv[n1]^2)` with `true`, the infinite loop is avoided, but I don't know if this is a proper thing to do.
https://github.com/JuliaLinearAlgebra/GenericLinearAlgebra.jl/blob/9f9b46f05df3eb13e3c889c64c9ed2ca72f450bb/src/svd.jl#L181-L188

**Comment by andreasnoack** [comment]
Thanks for reporting this. I can reproduce the issue. I'll post an update once I've figured out what is going on.

**Comment by andreasnoack** [comment]
https://github.com/JuliaLinearAlgebra/GenericLinearAlgebra.jl/pull/82 seems to fix the issue for the reported matrices here. Please let me know if you see any convergence problems for other matrices once the patch is released.

**Comment by andreasnoack** [comment]
This is indeed still an issue. I believe we need a dedicated solver for the 2x2 case.

**Comment by shinaoka** [comment]
Just for your information, I've faced a similar problem when trying to decompose a (ill-conditioned) Hilbert matrix using `Double64`. `svd` gets stuck in an infinite loop if the matrix size is larger than 16.

```
using DoubleFloats
using GenericLinearAlgebra

# Hilbert matrix
# N <= 16 is OK
N = 17
A = zeros(Double64, N, N)
for j in 1:N, i in 1:N
    A[i, j] = 1/Double64(i+j-1)
end

svd = GenericLinearAlgebra.svd(A)
```

**Comment by rabbott999** [comment]
I ran into a similar issue to the one described here and I believe I've found the issue. It looks like there's a typo at
https://github.com/JuliaLinearAlgebra/GenericLinearAlgebra.jl/blob/f3e9d27f83608f95b75773849e61c8d42bd0f7f4/src/svd.jl#L117 the line

``` μ = abs(dv[j + 1])*(μ*(μ + abs(ev[j]))) ```

should be

```μ = abs(dv[j + 1])*(μ/(μ + abs(ev[j])))```

This is what matches eq. 2.4 in http://www.netlib.org/lapack/lawnspdf/lawn03.pdf, and also the current version doesn't pass the check that scaling the matrix should scale μ by the same amount. If the bidiagonal matrix has small off-diagonal entries then μ can become very small (for the first matrix above matrix above I'm getting ~1e-41), which results in an impossible-to-meet stopping condition. Making the above change fixed all of the infinite loops I've found, including both of the matrices above.
